pkg/glusterfs/controllerserver.go

```diff
@@ -405,15 +454,243 @@ func (cs *ControllerServer) ControllerGetCapabilities(ctx context.Context, req *
+		return nil, status.Errorf(codes.Internal, "failed to activate snapshot %s", err.Error())
+	}
+	return &csi.CreateSnapshotResponse{
+		Snapshot: &csi.Snapshot{
+			Id:             snapResp.VolInfo.Name,
+			SourceVolumeId: snapResp.ParentVolName,
+			CreatedAt:      snapResp.CreatedAt.Unix(),
+			SizeBytes:      (int64(snapResp.VolInfo.Capacity)) * utils.MB,
+			Status: &csi.SnapshotStatus{
+				Type: csi.SnapshotStatus_READY,
+			},
+		},
+	}, nil
+}
+
+func (cs *ControllerServer) validateCreateSnapshotReq(req *csi.CreateSnapshotRequest) error {
+	if req == nil {
+		return status.Errorf(codes.InvalidArgument, "CreateSnapshot request is nil")
+	}
+	if req.GetName() == "" {
+		return status.Error(codes.InvalidArgument, "CreateSnapshot - name cannot be nil")
+	}
+
+	if req.GetSourceVolumeId() == "" {
+		return status.Error(codes.InvalidArgument, "CreateSnapshot - sourceVolumeId is nil")
+	}
+	if req.GetName() == req.GetSourceVolumeId() {
+		//In glusterd2 we cannot create a snapshot as same name as volume name
+		return status.Error(codes.InvalidArgument, "CreateSnapshot - sourceVolumeId  and snapshot name cannot be same")
+	}
+	return nil
 }
 
 // DeleteSnapshot delete provided snapshot of a PV
 func (cs *ControllerServer) DeleteSnapshot(ctx context.Context, req *csi.DeleteSnapshotRequest) (*csi.DeleteSnapshotResponse, error) {
-	return nil, status.Error(codes.Unimplemented, "")
+	if req == nil {
+		return nil, status.Errorf(codes.InvalidArgument, "DeleteSnapshot request is nil")
+	}
+	if req.GetSnapshotId() == "" {
+		return nil, status.Error(codes.InvalidArgument, "DeleteSnapshot - snapshotId is empty")
+	}
+	glog.V(4).Infof("deleting snapshot %s", req.GetSnapshotId())
+
+	err := cs.client.SnapshotDeactivate(req.GetSnapshotId())
+	if err != nil {
+		errResp := cs.client.LastErrorResponse()
+		if errResp != nil && errResp.StatusCode == http.StatusNotFound {
+			return &csi.DeleteSnapshotResponse{}, nil
+		}
+
+		if err.Error() != gd2Error.ErrSnapDeactivated.Error() {
+			glog.Errorf("failed to deactivate snapshot %v", err)
+			return nil, status.Errorf(codes.Internal, "DeleteSnapshot - failed to deactivate snapshot %s", err.Error())
+		}
+
+	}
+	err = cs.client.SnapshotDelete(req.SnapshotId)
+	if err != nil {
+		errResp := cs.client.LastErrorResponse()
+		if errResp != nil && errResp.StatusCode == http.StatusNotFound {
+			return &csi.DeleteSnapshotResponse{}, nil
+		}
+		glog.Errorf("failed to delete snapshot %v", err)
+		return nil, status.Errorf(codes.Internal, "DeleteSnapshot - failed to delete snapshot %s", err.Error())
+	}
+	return &csi.DeleteSnapshotResponse{}, nil
 }
 
 // ListSnapshots list the snapshots of a PV
 func (cs *ControllerServer) ListSnapshots(ctx context.Context, req *csi.ListSnapshotsRequest) (*csi.ListSnapshotsResponse, error) {
-	return nil, status.Error(codes.Unimplemented, "")
+	var (
+		snaplist   api.SnapListResp
+		err        error
+		startToken int32
+	)
+	if req.GetStartingToken() != "" {
```

Comment:
are we supporting pagination with the hack ? If not, we have to document atleast in the source.

Reply:
yes to make the CSI test cases to pass, this is a hack 
the issue is created in [here](https://github.com/kubernetes-csi/csi-test/issues/124)